specfile/specfile.py

```diff
@@ -514,3 +515,118 @@ def add_patch(
             index = patches.insert_numbered(number, location)
             if comment:
                 patches[index].comments.extend(comment.splitlines())
+
+    def update_value(
+        self, value: str, requested_value: str, protected_entities: Optional[str] = None
+    ) -> str:
+        """
+        Updates a value from within the context of the spec file with a new value,
+        but tries to preserve substitutions of locally defined macros and tags,
+        updating the respective macro definitions and tag values instead.
+
+        Args:
+            value: Value to update.
+            requested_value: Requested new value.
+            protected_entities: Regular expression specifying protected tags and macro definitions,
+              ensuring their values won't be updated.
+
+        Returns:
+            Updated value. Can be equal to the original value.
+        """
+        protected_regex = re.compile(
+            # (?!) doesn't match anything
+            protected_entities or "(?!)",
+            re.IGNORECASE,
+        )
+        # collect modifiable entities
+        entities = {}
+        with self.macro_definitions() as macro_definitions:
+            entities.update(
+                {
+                    md.name: {
+                        "value": md.body,
+                        "type": type(md),
+                        "locked": False,
+                        "changed": False,
+                    }
+                    for md in macro_definitions
+                    if not protected_regex.match(md.name)
+                    and not md.name.endswith(")")  # skip macro definitions with options
+                }
+            )
+        # order matters here - if there is a macro definition redefining a tag,
+        # we want to update the tag, not the macro definition
+        with self.tags() as tags:
+            entities.update(
+                {
+                    t.name.lower(): {
+                        "value": t.value,
+                        "type": type(t),
+                        "locked": False,
+                        "changed": False,
+                    }
+                    for t in tags
+                    if not protected_regex.match(t.name)
+                }
+            )
+        # tags can be referenced as %{tag} or %{TAG}
+        entities.update({k.upper(): v for k, v in entities.items() if v["type"] == Tag})
+
+        def update(value, requested_value):
+            regex, template = ValueParser.construct_regex(value, entities.keys())
+            m = regex.match(requested_value)
+            if m:
+                d = m.groupdict()
+                for grp, val in d.items():
+                    if grp.startswith("sub_"):
+                        continue
+                    if entities[grp]["locked"]:
+                        # avoid infinite recursion
+                        return requested_value
+                    entities[grp]["locked"] = True
```

Comment:
Small nitpick, the lock-unlock would be suited for a `try-finally` so that we always unlock even if an exception is thrown but it probably does not really matter here (if an exception arises, we are probably not handling it anywhere right now so no point in reseting the lock)

Reply:
Right, I decided to leave out `try-finally` in the end, but it's probably more future-proof.